### PR TITLE
Bottom Drawer New Feature: Skip Open State (API Break) + Crashes/Freezes Fixed!

### PR DESCRIPTION
**Pull request by Anubhvv**
### Description

1. We want to provide a new option in bottom drawer which allows the user to skip the Open State.
-      Close State --> Expanded State
-      Expanded State --> Close State 

2. When Expandable (or skipOpenState) param is changed, the BottomDrawer freezes/ Crashes on Open()/Expand() call.
3. BottomSliveOver BottomDrawer crashes when content inside is changed from More than Half Screen to Less than Half Screen, or vice versa.

### Root cause 

1.  Not Applicable, new feature
2. when exapndable argument's value is changed, the bottomDrawer still continues with old anchors and drawerState causing issues.
3. Target value not aligned with the old anchor values, so anchors need to be updated before they are used.

### Fix

1. So, Now one more parameter SkipOpenState (by default false) is used, and now Open state is similar to expanded state and how it behaves in case of expandable parameter.
2. Will make rememberDrawerState (as keys) & DrawerState accept arguments expandable and skipOpenState, so each time values are changed a new DrawerState is created and this approach is going to be in line with how Google Material Modal Bottom Sheet is handling this flow.
3. When the Content is changed from More than half height to less than half height and then we call Expand(), the expand sets target and call animateTo, the problem is new anchors are set much later in the process at bottomDrawerSwipeable(), and animateTo tries to use new Correct target Value with outdated anchors. 

### Validations

(how the change was tested, including both manual and automated tests)

### 1. skipOpenState

| Before                                       | After                                      |
|----------------------------------------------|--------------------------------------------|
|Screenshot_2023-09-29-00-08-46-41_ad3b258aa0cd3ec091b47fe7b71fec5b|IMG_20230929_001015|

https://github.com/microsoft/fluentui-android/assets/68989156/01fedb38-5c04-456c-a5dc-f51694fe47b1
`---------------------------------------------------------------------------------------------------------------------------`
### 2. Expandable toggling Crashes 

**Before**

https://github.com/microsoft/fluentui-android/assets/68989156/a0e10666-d3e9-43e8-a6b8-3888bed4c1ae

**After**

https://github.com/microsoft/fluentui-android/assets/68989156/2d762bf3-a421-4605-b2c6-3f8a8c2315ee

`---------------------------------------------------------------------------------------------------------------------------`
### 3. BottomSlideOver Full to half content 

**Before**

https://github.com/microsoft/fluentui-android/assets/68989156/af8381d9-83b4-4302-aa38-377d3f6b5a09

**After**

https://github.com/microsoft/fluentui-android/assets/68989156/8f171d0a-2ed0-46f0-9175-958612b885b5

### Pull request checklist

This PR has considered:
- [ ] Light and Dark appearances
- [x] Automated Tests
- [x] Documentation and demo app examples
- [x] VoiceOver and Keyboard Accessibility
- [ ] Internationalization and RTL layouts
- [ ] Size classes and window sizes (notched devices, multitasking, different window sizes, etc)

### CODE REVIEW

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 1]
```diff
@@ -931,6 +969,7 @@ fun BottomDrawer(
     expandable: Boolean = true,
     scrimVisible: Boolean = true,
     showHandle: Boolean = true,
+    skipOpenState: Boolean = false,
```
Comment (praveen970) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:972]: instead we can rename it to "swipeToClose". WDYT? skipOpenState can also be confused during the opening of drawer

Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:972]: Now, we have changed the functionality and it affects the opening as well.

File: FluentUI.Demo/src/main/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerActivity.kt [hunk 2]
```diff
@@ -265,12 +312,16 @@ private fun CreateActivityUI() {
...
     scrimVisible: Boolean,
     showHandle: Boolean,
     drawerContent: @Composable ((() -> Unit) -> Unit),
 ) {
+    val key = remember(slideOver, expandable, skipOpenState, scrimVisible, showHandle) {
+        "$slideOver-$expandable-$skipOpenState-$scrimVisible-$showHandle"
+    }
     val scope = rememberCoroutineScope()
-    val drawerState = rememberDrawerState()
+    val drawerState = remember(key) { DrawerState(initialValue = DrawerValue.Closed) }
```
Comment (mishramayank1) [inline FluentUI.Demo/src/main/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerActivity.kt:324]: why this change required.

File: fluentui_core/src/main/java/com/microsoft/fluentui/compose/Swipeable.kt
Comment (mishramayank1) [inline fluentui_core/src/main/java/com/microsoft/fluentui/compose/Swipeable.kt:1]: revert changes in this file.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 3]
```diff
@@ -126,15 +125,19 @@ class DrawerState(
         do{
             delay(50)
         }while(!anchorsFilled)
-
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:129]: update the function description

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 4]
```diff
@@ -172,7 +175,8 @@ class DrawerState(
 
         val targetValue = when {
             hasExpandedState -> DrawerValue.Expanded
-            else -> DrawerValue.Open
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:175]: update the function description

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 5]
```diff
@@ -767,12 +804,12 @@ private fun BottomDrawer(
...
                                         LocalContext.current.resources.getString(R.string.collapse)
                                     } else {
-                                        if (drawerState.hasExpandedState && drawerState.isOpen) LocalContext.current.resources.getString(
+                                        if (drawerState.hasExpandedState && !drawerState.isClosed) LocalContext.current.resources.getString(
                                             R.string.expand
                                         ) else null
                                     }
                                 ) {
-                                    if (drawerState.currentValue == DrawerValue.Expanded) {
+                                    if (drawerState.currentValue == DrawerValue.Expanded && drawerState.hasOpenedState) {
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:812]: put drawerState.hasOpenedState in below "if"

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 1 above)
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:972]: Update description

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 6]
```diff
@@ -647,7 +683,7 @@ private fun BottomDrawer(
             open = !drawerState.isClosed,
             onClose = onDismiss,
             fraction = {
-                if (drawerState.anchors.isEmpty()
+                if (drawerState.anchors.isEmpty() || skipOpenState
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:686]: Add if for hasExpandedState, hasOpenState to invoke calculate fraction respectively. If both of them false then return 0 to show invisible scrim.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 7]
```diff
@@ -706,7 +743,7 @@ private fun BottomDrawer(
                 .clip(drawerShape)
                 .background(drawerBackground)
                 .semantics {
-                    if (drawerState.isOpen) {
+                    if (drawerState.hasOpenedState && !drawerState.isClosed) {
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:746]: drawerState.hasOpenedState is not required

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 6 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:686]: I have added:
if( empty anchors)
0.dp
else if ( skipOpenState) calculate with EXPANDED
     else calcualte with OPEN

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 7 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:746]: Done

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 1 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:972]: Done

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 4 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:175]: done

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 3 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:129]: Done

File: fluentui_core/src/main/java/com/microsoft/fluentui/compose/Swipeable.kt
Comment (Anubhvv) [reply fluentui_core/src/main/java/com/microsoft/fluentui/compose/Swipeable.kt:1]: done

File: FluentUI.Demo/src/main/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerActivity.kt (hunk 2 above)
Comment (Anubhvv) [reply FluentUI.Demo/src/main/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerActivity.kt]: removed key part

File: FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt [hunk 8]
```diff
@@ -241,4 +248,18 @@ class V2BottomDrawerUITest : BaseTest() {
         }
         closeCheckForVerticalDrawer()
     }
+    @Test
+    fun testNonExpandableAndExpandDrawerClicked() {
+        // Set expandable = false
+        composeTestRule.onNodeWithText("Expandable", useUnmergedTree = true).performClick()
```
Comment (mishramayank1) [inline FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt:254]: Use String Id to retrieve string. This is required to run test in other locale.

Comment (Anubhvv) [reply FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt]: Done.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 9]
```diff
@@ -541,7 +630,7 @@ private fun TopDrawer(
                 drawerConstraints
                     .offset { IntOffset(0, 0) }
                     .semantics {
-                        if (drawerState.isOpen) {
+                        if (drawerState.hasOpenedState && !drawerState.isClosed) {
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:633]: not required "drawerState.hasOpenedState"

File: FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt [hunk 10]
```diff
@@ -251,7 +251,7 @@ class V2BottomDrawerUITest : BaseTest() {
     @Test
     fun testNonExpandableAndExpandDrawerClicked() {
         // Set expandable = false
-        composeTestRule.onNodeWithText("Expandable", useUnmergedTree = true).performClick()
+        composeTestRule.onNodeWithText(getString(R.string.drawer_expandable), useUnmergedTree = true).performClick()
         // Click on the "Open Drawer" button
         composeTestRule.onNodeWithText("Expand Drawer").performClick()
         //Skip Open State Button must be disabled
```
Comment (mishramayank1) [inline FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt:257]: Use string Id at other places as well.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 11]
```diff
@@ -123,20 +138,34 @@ class DrawerState(
...
+        do {
             delay(50)
-        }while(!anchorsFilled)
+        } while (!anchorsFilled)
+        /*
+        * open and expanded states are possible in case of
+        * true expandanble and false skipOpenState respectively.
+         */
+        var targetValue = when {
+            hasOpenedState -> DrawerValue.Open
```
[the remark this answers is not in the document]
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:149]: okay

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 12]
```diff
@@ -166,33 +197,60 @@ class DrawerState(
...
             delay(50)
-        }while(!anchorsFilled)
-
+        } while (!anchorsFilled)
+        /*
+        * expandable and open states are possible in case of
+        * true expandanble and false skipOpenState respectively.
+         */
         val targetValue = when {
             hasExpandedState -> DrawerValue.Expanded
```
[the remark this answers is not in the document]
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:208]: okay

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 13]
```diff
@@ -428,7 +517,7 @@ private fun HorizontalDrawer(
                         end = if (leftSlide) paddingPx else 0.dp
                     )
                     .semantics {
-                        if (drawerState.isOpen) {
+                        if (drawerState.hasOpenedState && !drawerState.isClosed) {
```
[the remark this answers is not in the document]
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:520]: okay, removed  "drawerState.hasOpenedState"

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 9 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:633]: okay, removed "drawerState.hasOpenedState"

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 14]
```diff
@@ -623,6 +712,7 @@ private fun BottomDrawer(
     scrimColor: Color,
     scrimVisible: Boolean,
     expandable: Boolean,
+    skipOpenState: Boolean,
```
[the remark this answers is not in the document]
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:715]: Done!

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 15]
```diff
@@ -647,15 +737,29 @@ private fun BottomDrawer(
...
-                    )
+                    if(skipOpenState){
+                        calculateFraction(
+                            drawerState.anchors.entries.firstOrNull { it.value == DrawerValue.Closed }?.key!!,
+                            drawerState.anchors.entries.firstOrNull { it.value == DrawerValue.Expanded }?.key!!,
+                            drawerState.offset.value
+                        )
+                    }else{
+
+                        if(drawerState.anchors.entries.firstOrNull { it.value == DrawerValue.Open }?.key != null) {
```
[the remark this answers is not in the document]
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:752]: Yes, I have changed the logic.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 16]
```diff
@@ -978,6 +1087,7 @@ fun BottomDrawer(
                 scrimColor = scrimColor,
                 scrimVisible = scrimVisible,
                 expandable = expandable,
+                skipOpenState = skipOpenState,
```
[the remark this answers is not in the document]
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:1090]: done

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 17]
```diff
@@ -913,9 +1020,10 @@ fun Drawer(
  * @param slideOver if true, then BottomDrawer would be drawn in full length & only covering up to half screen when open & it get slided more
  * in the visible region on expand. If false then, the BottomDrawer end at the bottom & hence the content get only the visible region height to draw itself.The default value is true
  * @param expandable if true drawer would expand on drag else drawer open till fixed/wrapped height.
- * The default value is true
+ * The default value is true. If expandable is
```
[the remark this answers is not in the document]
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:1023]: And the description has been moved to  DrawerState part

File: FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt (hunk 10 above)
Comment (Anubhvv) [reply FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt:257]: Got it, done

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 18]
```diff
@@ -80,24 +80,41 @@ enum class DrawerValue {
...
  *
  * @param initialValue The initial value of the state.
+ * @param expandable defines if the drawer is allowed to take the Expanded state.
+ * @param isSkipOpenState defines if the drawer is allowed to take the Open state. (Open State is skipped in case of true)
  * @param confirmStateChange Optional callback invoked to confirm or veto a pending state change.
  */
 class DrawerState(
     private val initialValue: DrawerValue = DrawerValue.Closed,
+    internal val expandable: Boolean = true,
+    internal val isSkipOpenState: Boolean = false,
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:90]: rename to "skipOpenState" & update description as well & other places.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 19]
```diff
@@ -80,24 +80,41 @@ enum class DrawerValue {
...
+        if (isSkipOpenState) {
+            require(initialValue != DrawerValue.Open) {
+                "The initial value must not be set to Open if skipOpenState is set to" +
+                        " true."
+            }
+        }
+        else if (expandable) {
+            require(initialValue != DrawerValue.Expanded) {
+                "The initial value must not be set to Expanded if expandable is set to" +
+                        " true."
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:107]: instead of "true" here. It should be false ?

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 20]
```diff
@@ -80,24 +80,41 @@ enum class DrawerValue {
...
 ) {
-
+    init {
+        if (isSkipOpenState) {
+            require(initialValue != DrawerValue.Open) {
+                "The initial value must not be set to Open if skipOpenState is set to" +
+                        " true."
+            }
+        }
+        else if (expandable) {
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:104]: remove the else. check for expandable is independent check

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 21]
```diff
@@ -123,20 +140,35 @@ class DrawerState(
     suspend fun open() {
         enable = true
         animationInProgress = true
-        do{
+        do {
             delay(50)
-        }while(!anchorsFilled)
+        } while (!anchorsFilled)
+        /*
+        * open and expanded states are possible in case of
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:147]: rephrase the comment for clarity

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 22]
```diff
@@ -166,33 +200,61 @@ class DrawerState(
...
         enable = true
         animationInProgress = true
-        do{
+        do {
             delay(50)
-        }while(!anchorsFilled)
-
+        } while (!anchorsFilled)
+        /*
+        * expandable and open states are possible in case of
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:207]: rephrase the comment for clarity

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 23]
```diff
@@ -166,33 +200,61 @@ class DrawerState(
...
+                animationInProgress = false
+            }
+        } else {
             animationInProgress = false
-        } catch (e: Exception) {
-            animateTo(targetValue = targetValue, AnimationSpec)
         }
     }
 
+
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:228]: remove extra line

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 24]
```diff
@@ -315,17 +391,26 @@ private fun Modifier.bottomDrawerSwipeable(
...
+                    maxOpenHeight to DrawerValue.Open,
+                    0F to DrawerValue.Expanded
+                )
+            }
         } else {
             mapOf(
                 fullHeight to DrawerValue.Closed,
                 maxOpenHeight to DrawerValue.Open
             )
+
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:413]: remove extra line

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 25]
```diff
@@ -202,11 +264,18 @@ class DrawerState(
...
 @Composable
 fun rememberDrawerState(confirmStateChange: (DrawerValue) -> Boolean = { true }): DrawerState {
-    return rememberSaveable(saver = DrawerState.Saver(confirmStateChange)) {
-        DrawerState(DrawerValue.Closed, confirmStateChange)
+    return rememberSaveable(saver = DrawerState.Saver(expandable = true, skipOpenState = false, confirmStateChange = confirmStateChange)) {
+        DrawerState(initialValue = DrawerValue.Closed, expandable = true, isSkipOpenState = false, confirmStateChange)
+    }
+}
+@Composable
+fun rememberBottomDrawerState(initialValue: DrawerValue = DrawerValue.Closed, confirmStateChange: (DrawerValue) -> Boolean = { true }, expandable: Boolean = true, skipOpenState: Boolean = false ): DrawerState {
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:272]: put confirmStateChange param at the end

Comment (mishramayank1) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:272]: don't expose "initialValue". It would throw error if user set it to expanded state but expanded state is not available for content size having less than half of screen.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 19 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:107]: Thanks for pointing that out, fixed.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 20 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:104]: Done.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 21 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:147]: I have changed comments, and descriptions for better clarity.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 22 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:207]: Done.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 25 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:272]: I have unexposed initialValue, and put confirmStateChange param at the end.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 26]
```diff
@@ -292,17 +366,25 @@ private fun Modifier.bottomDrawerSwipeable(
...
-                    bottomExpandedStateY to DrawerValue.Expanded
-                )
+                if(drawerState.expandable){
+                    mapOf(
+                        fullHeight to DrawerValue.Closed,
+                        bottomOpenStateY to DrawerValue.Open,
+                        bottomExpandedStateY to DrawerValue.Expanded
+                    )
+                }
+                else {
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:382]: else part wont be required as per above comment

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 27]
```diff
@@ -647,10 +736,10 @@ private fun BottomDrawer(
             open = !drawerState.isClosed,
             onClose = onDismiss,
             fraction = {
-                if (drawerState.anchors.isEmpty()
-                ) {
+                if (drawerState.anchors.isEmpty() || drawerState.skipOpenState) {
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:739]: scrim need to be shown when the anchor is calculated. Remove the  drawerState.skipOpenState

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 28]
```diff
@@ -647,10 +736,10 @@ private fun BottomDrawer(
...
                     0.toFloat()
-                } else {
+                }
+                else {
                     calculateFraction(
                         drawerState.anchors.entries.firstOrNull { it.value == DrawerValue.Closed }?.key!!,
                         drawerState.anchors.entries.firstOrNull { it.value == DrawerValue.Open }?.key!!,
```
Comment (mishramayank1) [inline fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:745]: Get the second parameter as per the next state from close based on the parameter provided in DrawerState

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 29]
```diff
@@ -292,17 +366,25 @@ private fun Modifier.bottomDrawerSwipeable(
             val minHeight = 0f
             val bottomOpenStateY = max(maxOpenHeight, fullHeight - drawerHeight)
             val bottomExpandedStateY = max(minHeight, fullHeight - drawerHeight)
-            val anchors = if (drawerHeight <= maxOpenHeight || !expandable) {
+            val anchors = if (drawerHeight <= maxOpenHeight || drawerState.skipOpenState) {
```
[the remark this answers is not in the document]
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:369]: Oh, right, i think the check should be !drawerState.skipOpenState bec we are adding Open state in anchors

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 27 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:739]: If you look at the else part, it searches for the anchor value corresponding to the OPEN state, and first time after skipOpenState = true has been set when this is being executed, anchors haven't been created yet, that throws NULL POINTER error.

Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:739]: Okay, will handle that case in else part

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 26 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:382]: now this else contains check for skipOpenState and passes the second parameter accordingly.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 28 above)
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:745]: okayy

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt [hunk 30]
```diff
@@ -701,12 +789,18 @@ private fun BottomDrawer(
                         }
                     }
                 }
-                .drawerHeight(expandable, slideOver, maxOpenHeight, fullHeight, drawerState)
+                .drawerHeight(
+                    drawerState.hasExpandedState,
```
[the remark this answers is not in the document]
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt:793]: Thanks for pointing this out.

File: fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt (hunk 29 above)
[the remark this answers is not in the document]
Comment (Anubhvv) [reply fluentui_drawer/src/main/java/com/microsoft/fluentui/tokenized/drawer/Drawer.kt]: Removed the check

File: FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt [hunk 31]
```diff
@@ -241,4 +242,18 @@ class V2BottomDrawerUITest : BaseTest() {
         }
         closeCheckForVerticalDrawer()
     }
+    @Test
+    fun testNonExpandableAndExpandDrawerClicked() {
+        // Set expandable = false
+        composeTestRule.onNodeWithText(getString(R.string.drawer_expandable), useUnmergedTree = true).performClick()
+        // Click on the "Open Drawer" button
+        composeTestRule.onNodeWithText(getString(R.string.drawer_expand)).performClick()
```
Comment (mishramayank1) [inline FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt:250]: this is called below as well. remove it.

File: FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt [hunk 32]
```diff
@@ -241,4 +242,18 @@ class V2BottomDrawerUITest : BaseTest() {
         }
         closeCheckForVerticalDrawer()
     }
+    @Test
+    fun testNonExpandableAndExpandDrawerClicked() {
```
Comment (mishramayank1) [inline FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt:246]: rename to testSetNonExpandableThenExpandableDrawerClicked

File: FluentUI.Demo/src/main/res/values-night/strings.xml [hunk 33]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (mishramayank1) [inline FluentUI.Demo/src/main/res/values-night/strings.xml:1]: why this is required ?

File: FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt (hunk 31 above)
Comment (mishramayank1) [reply FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt:250]: Should remove this line instead of below.

Comment (Anubhvv) [reply FluentUI.Demo/src/androidTest/java/com/microsoft/fluentuidemo/demos/V2BottomDrawerUITest.kt]: Done

File: FluentUI.Demo/src/main/res/values-night/strings.xml (hunk 33 above)
Comment (Anubhvv) [reply FluentUI.Demo/src/main/res/values-night/strings.xml]: removed this file

**Review by mishramayank1 — APPROVED** [review]
